crates/kitsune_p2p/dht_arc/src/test/arc_stability.rs

```diff
@@ -1,104 +1,293 @@
-/// achieved.
-const STEADY_WINDOW: usize = 3;
+/// Number of consecutive rounds of no movement before declaring convergence.
+const CONVERGENCE_WINDOW: usize = 3;
 
+/// Level of detail in reporting.
 const DETAIL: u8 = 0;
 
+type DataVec = statrs::statistics::Data<Vec<f64>>;
+
+pub type Peers = Vec<DhtArc>;
+
+fn max_halflen() -> f64 {
+    MAX_HALF_LENGTH as f64
+}
+
 fn full_len() -> f64 {
     2f64.powi(32)
 }
 
+pub fn seeded_rng(seed: Option<u64>) -> StdRng {
+    let seed = seed.unwrap_or_else(|| thread_rng().gen());
+    tracing::info!("RNG seed: {}", seed);
+    StdRng::seed_from_u64(seed)
+}
+
 #[test]
-fn min_redundancy_is_maintained() {
-    todo!("Check that min redundancy is maintained at all times");
+fn single_agent_convergence_debug() {
+    std::env::set_var("RUST_LOG", "debug");
+    observability::test_run().ok();
+
+    let n = 50;
+    let j = 0.1;
+    let redundancy = 5;
+    let check_gaps = false;
+
+    let mut rng = seeded_rng(None);
+    // let mut rng = seeded_rng(Some(5181023930453438019));
+
+    let strat = PeerStratAlpha {
+        check_gaps,
+        redundancy_target: redundancy / 2,
+        ..Default::default()
+    }
+    .into();
+
+    let s = ArcLenStrategy::Constant(redundancy as f64 / n as f64);
+
+    let mut peers = simple_parameterized_generator(&mut rng, n, j, s);
+    peers[0].half_length = MAX_HALF_LENGTH;
+    tracing::debug!("{}", EpochStats::oneline_header());
+    let runs = determine_equilibrium(1, peers, |peers| {
+        let dynamic = Some(maplit::hashset![0]);
+        let (peers, stats) = run_one_epoch(&strat, peers, dynamic.as_ref(), DETAIL);
+
+        tracing::debug!("{}", stats.oneline());
+        // tracing::debug!("{}", peers[0].coverage());
+        (peers, stats)
+    });
+    print_arcs(&runs.0[0].peers);
+    report(&runs);
 }
 
+#[allow(dead_code)]
+fn run_single_agent_convergence(
+    iters: usize,
+    n: usize,
+    redundancy: u16,
+    j: f64,
+    check_gaps: bool,
+) -> RunBatch {
+    tracing::info!("");
+    tracing::info!("------------------------");
+
+    // let seed = None;
+    let seed = Some(7532095396949412554);
+    let mut rng = seeded_rng(seed);
+
+    let strat = PeerStratAlpha {
+        check_gaps,
+        redundancy_target: redundancy / 2,
+        ..Default::default()
+    }
+    .into();
+
+    let s = ArcLenStrategy::Constant(redundancy as f64 / n as f64);
+
+    let mut peers = simple_parameterized_generator(&mut rng, n, j, s);
+    peers[0].half_length = MAX_HALF_LENGTH;
+    let runs = determine_equilibrium(iters, peers, |peers| {
+        let dynamic = Some(maplit::hashset![0]);
+        let (peers, stats) = run_one_epoch(&strat, peers, dynamic.as_ref(), DETAIL);
+        tracing::debug!("{}", peers[0].coverage());
+        (peers, stats)
+    });
+    report(&runs);
+    runs
+}
+
+/// Test if various distributions of agents can converge
+#[test]
+#[cfg(feature = "slow_tests")]
```

Comment:
Just fyi slow_tests is enabled in CI, so if these are really really slow you might want to ignore them for CI.